### PR TITLE
Added support for msearch API to pass search pipeline name

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
Added support for msearch API to pass search pipeline name.
With this change a search pipeline name can be provided while making a msearch call
```
{ "index": "test"}
{ "query": { "match_all": {} }, "from": 0, "size": 10, "search_pipeline": "my_pipeline"}
{ "index": "test-1", "search_type": "dfs_query_then_fetch"}
{ "query": { "match_all": {} }, "search_pipeline": "my_pipeline1" }
```

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/15748

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable. - https://github.com/opensearch-project/documentation-website/pull/8372

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by epugh** [comment]
Exciting!  A couple of comments/questions:

1) Can I use different pipelines per query?  Or just one pipeline?   If I can't use different pipelines, and I try, do I get a nice error ;-).    I can see future Eric trying things that make life hard!

2) I sort of wish the same approach applied to the normal `_search` end point as well.   Is there any overlap/refactoring if you did both at the same time?   Or are the codes bases pretty different.

This will totally scratch my itch!   Hoping it makes it in very soon!

**Comment by dblock** [comment]
Please don't forget https://github.com/opensearch-project/opensearch-api-specification!

**Comment by owaiskazi19** [comment]
>  1. Can I use different pipelines per query?  Or just one pipeline?   If I can't use different pipelines, and I try, do I get a nice error ;-).    I can see future Eric trying things that make life hard!

Yes, we should be able to use 2 different pipelines for a request as they are independent of each other.

```json
{ "index": "test"}
{ "query": { "match_all": {} }, "from": 0, "size": 10, "search_pipeline": "my_pipeline"}
{ "index": "test-1", "search_type": "dfs_query_then_fetch"}
{ "query": { "match_all": {} }, "search_pipeline": "my_pipeline1" }

```
> 2. I sort of wish the same approach applied to the normal `_search` end point as well.   Is there any overlap/refactoring if you did both at the same time?   Or are the codes bases pretty different.

Yes, tried the same with `_search` and works well

```json
{
    "query": {
        "match_all": {}
    },
    "from": 0,
    "size": 10,
    "search_pipeline": "my_pipeline1"
}
```

File: server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:400]: Rather than checking both properties, I think I would suggest making sure the `SearchRequest` and `SearchSourceBuilder` are in agreement. If you check `org.opensearch.rest.action.search.RestSearchAction#parseSearchSource`, you can see how URL parameters get pushed down into the `SearchSourceBuilder`. I would apply something similar here.

Alternatively, you could "lift" the value up into `SearchRequest` for backward compatibility. In `org.opensearch.rest.action.search.RestSearchAction#parseSearchRequest`, after the call to `parseSearchSource`, I think you could change [this line](https://github.com/opensearch-project/OpenSearch/blob/ab7e914c3956ac206cf269ca345574882770a283/server/src/main/java/org/opensearch/rest/action/search/RestSearchAction.java#L213) to:
```
searchRequest.pipeline(request.param("search_pipeline", searchRequest.source().pipeline()));
```

I think I like that second option more.

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:400]: Of course, unit testing it is mildly annoying. We don't have any unit tests for `RestSearchAction`. 

Adding test coverage for `parseSearchRequest` would be helpful, though.

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:400]: @msfroh wouldn't the rest request call of RestMultiSearchAction would go through the transport layer of SearchRequest [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L320)? This will not allow to read the params on the rest side of SearchAction. Wdyt?

Comment (msfroh) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java:400]: Oh -- you're right. The `SearchRequest` objects for `msearch` are constructed in [here](https://github.com/opensearch-project/OpenSearch/blob/6823ca1291ddcfc63d0704f93990560af1731e3f/server/src/main/java/org/opensearch/action/search/MultiSearchRequest.java#L224).

In theory, you could use the above line to lift the value from source into `SearchRequest` for `SearchAction` and `SearchTemplateAction`. Somewhere around [this line](https://github.com/opensearch-project/OpenSearch/blob/6823ca1291ddcfc63d0704f93990560af1731e3f/server/src/main/java/org/opensearch/action/search/MultiSearchRequest.java#L313) you could lift the value into `SearchRequest` from the `SearchSourceBuilder` as well.

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/pipeline/SearchPipelineService.java]: This looks like a better place. I have pulled the change in `MultiSearchRequest.java`

**Review by msfroh — COMMENTED** [review]
Ahh! You lost the part that copies from the source into `SearchRequest` on the regular `_search` path. It would be really nice to get that part too.

File: server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:353]: You'll need to guard serialization and deserialization of this field with version checks.

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:353]: Was just about to push that :)

File: server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java
Comment (msfroh) [inline server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java:304]: You'll need to do the whole BWC dance, since 2.x is building 2.18.

1. Do this on main with `onOrAfter(Version.V_3_0_0))`.
2. You'll need a manual backport to 2.x, where you do `onOrAfter(Version.V_2_18_0)`.
3. Before merging the backport to 2.x, open another PR on main to change it to `onOrAfter(Version.V_2_18_0)`.
4. Merge the backport PR.
5. Merge the main version update PR.

We need to do those last two steps as close together as possible, as merging the backport PR will break builds on main.

I hate this process, but we haven't been able to think of something better.

Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/search/builder/SearchSourceBuilder.java]: Yes, you are right. Been a while changed an API body. Pushed the change and also added one more test for serialization.

**Comment by owaiskazi19** [comment]
@msfroh can I get another review on this?

**Review by msfroh — APPROVED** [review]
Nice! Thanks, @owaiskazi19!

**Comment by owaiskazi19** [comment]
@msfroh 
2.x PR: https://github.com/opensearch-project/OpenSearch/pull/16085
BWC version change PR to main: https://github.com/opensearch-project/OpenSearch/pull/16086

**Comment by msfroh** [comment]
> @msfroh 
> 
> 2.x PR: https://github.com/opensearch-project/OpenSearch/pull/16085
> 
> BWC version change PR to main: https://github.com/opensearch-project/OpenSearch/pull/16086

@owaiskazi19 I'll merge both of these tomorrow morning. I've made the mistake of merging the first PR then going to sleep a few too many times. 😀

**Comment by dblock** [comment]
This is an API change, isn't it? It needs changes in api-specification repo please?

**Pull request merged**
